README
Perl-Dist-WiX version 1.200002

DESCRIPTION
    This package is the upgrade to Perl::Dist based on Windows Installer XML
    technology, instead of Inno Setup.

    Perl distributions built with this module have the option of being
    created as Windows Installer databases (otherwise known as .msi files)

INSTALLATION
    To install this module, run the following commands:

            perl Makefile.PL
            make
            make test
            make install

DEPENDENCIES
    Perl 5.8.1 is the mimimum version of perl that this module will run on.

    Other modules that this module depends on are a working version of
    Alien::WiX, Data::Dump::Streamer 2.08, Data::UUID 1.149,
    Devel::StackTrace 1.20, Exception::Class 1.22, File::ShareDir 1.00,
    IO::String 1.08, List::MoreUtils 0.07, Module::CoreList 2.29, Win32::Exe
    0.13, Object::InsideOut 3.53, Perl::Dist 1.14, Process 0.26, Readonly
    1.03, URI 1.35, and Win32 0.35.

COPYRIGHT AND LICENSE
    Copyright 2009 - 2010 Curtis Jewell.

    Copyright 2008 - 2009 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Changes for Perl extension Perl-Dist-WiX

1.200002   Wed 02 Jun 2010
    1. Used to build Strawberry Perl + Padre 0.62.
    2. New version of Moose revealed some bugs. Fixing.

1.200001   Thu 06 May 2010
    1. Used to build Bootstrap Perl "April" 2010.
    2. Used to build Strawberry Perl 5.12.0.1.
    3. Requires version of Win32::Exe that has 64-bit support.
       (should have been in 1.200)
    4. Fixing small bugs in checkpointing.

1.200   Sat 01 May 2010
    1. Documentation cleanup (including adding when required).
    2. Supports 5.12.0 final. Removes 5.11.5 support.
    3. Used to build Strawberry Perl April 2010.
    4. Checkpointing works again.
    5. .tar.bz2 files are now usable as parameters to this module.
    6. (beta, untested) .tar.xz file support has also been added.
    7. A relocation custom action has been added.

1.102_102   Thu 25 Mar 2010
    1. Adding 5.12.0 support (based on 5.12.0-RC0 for now).
    2. Bug fixes to relocation support.
    3. Used to build Strawberry Perl April 2010 Beta 1
       (5.12.x versions).

1.102_101   Sat 20 Mar 2010
    1. Scripts now will go to perl/site/bin.
    2. perl 5.11.5 support is added.
    3. Relocation script written in perl has been added.
    4. Change $Config{installsitebin} from perl/bin to perl/site/bin,
       and add perl/site/bin to the path.
    5. Used to build Strawberry Perl April 2010 Beta 1
       (5.8.x/5.10.x versions).

1.102_100   Fri 19 Feb 2010
    1. Used to build Strawberry Perl Professional 5.10.1.1 Alpha 1.
    2. Allow more than one DISTRIBUTIONS.txt file.

1.102002   Sat 30 Jan 2010
    1. Allow use of previously created merge modules.
    2. Removal of MooseX::AttributeHelpers as a
       prerequisite.

1.102001   Fri 29 Jan 2010
    1. Changes to get Bootstrap to build.

1.102   Thu 28 Jan 2010
    1. Used to build Strawberry Perl January 2010.
    2. Fix problems with merge module
    3. Make updating "smart" - don't delete the site
       directory unless doing a major upgrade.
    4. Bring Portable back (install all modules into
       "perl/site" for it only.)
    5. Work on "gcc-4" toolchains (not done yet)
    6. Change to building with Module::Build.	

1.101_001   Wed 11 Nov 2009
    1. Split most of the build into a merge module.
    2. Finished the Moosification.
    3. Added ability to build off a git checkout.
    4. Started work on documentation. 

1.100   Wed 11 Nov 2009
    1. Used to make Strawberry Perl October 2009.
    2. Used to make Padre Standalone 0.50.
    3. Small fixes (adding prerequisite that _103 missed,
       removing reordering [had to take out portable for now]
       and making sure root directory files got installed.)

1.090_103   Sat 26 Sep 2009
    1. Used to make Strawberry Perl October 2009 Beta 2.
    2. Portability now works.
    3. Small bug fixes.

1.090_102   Mon 14 Sep 2009
    1. Major refactoring. (Now uses Moose rather 
       than Object::InsideOut.)
    2. 5.10.1 is now supported.
    3. Perl::Dist is not a dependency anymore.
    4. Perl::Dist::WiX now installs its non-core 
       modules into "perl\vendor" as opposed to 
       "perl\site"
       
1.000   Sat 25 Jul 2009.
    1. Used to make "Padre Standalone Plus Six" for the
       Padre Birthday party.
    2. Major version number jump - this is stable enough
       to not be a 0.x version anymore. 
    2. build_number can now be 0.
    3. minicpan script updated to take command-line options
       in order to be able to make minicpans for subclasses.
    4. minicpan script is more "PBP-compliant" (i.e. 
       the standard meta-options like --help are there)
    5. The cpan/sources directory is moved to the side when 
       installing from a minicpan.
    6. 'six' is now correctly integrated. (Note that 'six'
       is not guaranteed to be stable or usable.)

0.192   Fri 17 Jul 2009.
    1. Will be used to make Strawberry Perl July 2009.
    2. Windows XP is now the minimum supported Windows 
       version.
    3. Installs Win32::Process and IPC::System::Simple 
       if autodie is installed.
    4. Clears PAR temporary directory.
    5. Installs requirements for portability individually.
    6. First attempt at Perl 6 support.
	
0.191   Mon 06 Jul 2009
    1. Was used to make Padre Standalone beta 1.
    2. Tries to get a DefaultLanguage code for files 
       that require it.
    3. Moving the default fragments directory location under 
       temp_dir rather than output_dir.
    4. Stop CPAN.pm from following prerequisites during
       install_module.
    5. Updated CPANPLUS::Config template.   

0.190   Thu 02 Jul 2009
    1. Was used to make Strawberry Perl July 2009 
       Beta 3.
    2. Throw an exception if Archive::Zip version 1.28 
       will be installed instead. (Forcing 1.26 did not work.)
    3. A CPANPLUS::Config has been created that does similar
       things to the CPAN::Config that is installed.
    4. Building completely offline can now be successfully done.
       Because of this, CPAN::SQLite is disabled when actually
       installing modules during distribution building.
    5. Will attempt to tell you which directories need to be 
       added to the msi_directory_tree_additions parameter to 
	   Perl::Dist::WiX->new when trace is 2 or greater.
    6. Added a distropref so that certain modules will not
       be installed when building a distribution. (It is not 
       installed with the created distribution.)

0.185   Tue 23 Jun 2009
    1. Force install of Archive::Zip version 1.26 (1.28 
       has a bug in it.)
    2. Builds a list of CPAN distributions that were 
       installed as DISTRIBUTIONS.txt in the root directory
       of the install.
    3. A few more files are removed.
    4. Checks of CPAN use the CPAN passed in, not the 
       CPAN of the host perl or cpan.strawberryperl.com.
    5. Required version of Perl::Dist increased to 1.14 
       (except for Perl::Dist::Util::Toolchain, which has
	   to be gotten from SVN for now.)


Build.PL
use inc::latest Module::Build;

# CPAN version test.
require CPAN;
if ($CPAN::VERSION < 1.9457) {
	print <<'EOF';
To install the rest of the prerequisites, you need to run

	cpan ANDK/CPAN-1.94_57.tar.gz

before installing this. 

If you're already in CPAN or CPANPLUS, replace 'cpan' with 'install'.
If you're in CPAN, 'reload cpan' afterwards.

EOF
	die "Please upgrade CPAN as instructed above.\n\n";
}

die "OS unsupported\n" unless 'MSWin32' eq $^O;

# These are the main requirements.
my $requires = {
	'perl'                      => '5.008001',
	'parent'                    => '0.221',
	'Alien::WiX'                => '0.300000',
	'Archive::Tar'              => '1.42',
	'Archive::Zip'              => '1.26',
	'CPAN'                      => '1.9457',
	'Clone'                     => '0.31',
	'Data::Dump::Streamer'      => '2.08',
	'Data::UUID'                => '1.149',
	'Digest::CRC'               => '0.14',
	'Devel::StackTrace'         => '1.20',
	'Exception::Class'          => '1.22',
	'File::Copy::Recursive'     => '0.38',
	'File::Find::Rule'          => '0.30',
	'File::HomeDir'             => '0.82',
	'File::List::Object'        => '0.201',
	'File::Path'                => '2.07',
	'File::PathList'            => '1.04',
	'File::pushd'               => '1.00',
	'File::Remove'              => '1.42',
	'File::ShareDir'            => '1.00',
	'File::Spec'                => '3.29',
	'File::Temp'                => '0.21',
	'HTTP::Status'              => '5.817',
	'IPC::Run3'                 => '0.042',
	'IO::Capture'               => '0.05',
	'IO::String'                => '1.08',
	'List::MoreUtils'           => '0.07',
	'LWP::UserAgent'            => '5.823',
	'LWP::UserAgent::WithCache' => '0.06',
	'LWP::Online'               => '1.07',
	'Module::CoreList'          => '2.29',
	'Moose'                     => '0.90',
	'MooseX::NonMoose'          => '0.05',
	'MooseX::Singleton'         => '0',
	'MooseX::Types'             => '0',
	'MooseX::Types::URI'        => '0.02',
	'MooseX::Types::Path::Class' => '0.05',
	'namespace::autoclean'      => '0.10',
	'PAR::Dist'                 => '0.42',
	'Params::Util'              => '0.35',
	'Portable::Dist'            => '0.03',
	'Probe::Perl'               => '0.01',
	'Process'                   => '0.26',
	'Readonly'                  => '1.03',
	# Template 2.21 will not install on Windows, anyway.
	'Template'                  => '2.22',
	'URI'                       => '1.35',
	'WiX3'                      => '0.009100',
	'YAML::Tiny'                => '1.36',
	'Win32'                     => '0.35',
	'Win32::Exe'                => '0.13',
	'Win32::File::Object'       => '0.01',
	'Win32API::File'            => '0.1001',
};

my $class = Module::Build->subclass(
	class => 'PDWiX::Builder',
	code  => <<'END_CODE',
sub ACTION_authortest {
    my ($self) = @_;

    $self->depends_on('build');
    $self->depends_on('manifest');
    $self->depends_on('distmeta');

    $self->test_files( qw< t xt > );
    $self->depends_on('test');

    return;
}


sub ACTION_releasetest {
    my ($self) = @_;

    $self->depends_on('build');
    $self->depends_on('manifest');
    $self->depends_on('distmeta');

    $self->test_files( qw< t xt rt > );
    $self->depends_on('test');

    return;
}


sub ACTION_manifest {
    my ($self, @arguments) = @_;

    if (-e 'MANIFEST') {
        unlink 'MANIFEST' or die "Can't unlink MANIFEST: $!";
    }

    return $self->SUPER::ACTION_manifest(@arguments);
}
END_CODE
);


my $builder = $class->new(
	module_name         => 'Perl::Dist::WiX',
	license             => 'perl',
	dist_author         => 'Curtis Jewell <csjewell@cpan.org>',
	dist_version_from   => 'lib/Perl/Dist/WiX.pm',
	create_makefile_pl  => 'small',
	create_license      => 1,
	create_readme       => 1,
# Will do this soon, but I have to get a key first.
#	sign                => 1,
	configure_requires => {
		'Module::Build'             => '0.36',
		'CPAN'                      => '1.9456',
	},
	requires => $requires,
	build_requires => {
		'Test::LongString'          => '0.11',
		'Test::More'                => '0.86',
		'Test::Output'              => '0.11',
		'Test::Perl::Dist'          => '0.300',
		'Test::Script'              => '1.03',
		'Test::UseAllModules'       => '0.09',
	},
	recommends => {
		'Readonly::XS'              => '1.05',
	},
	meta_add => {
		no_index => {
			directory => [qw(inc share t)],
		},
	}, 
	meta_merge => {
		resources => {
			homepage    => 'http://csjewell.comyr.com/perl/',
			bugtracker  => 'http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Perl-Dist-WiX',
			repository  => 'http://svn.ali.as/cpan/trunk/Perl-Dist-WiX/',
			IRC         => 'irc://irc.perl.org/#win32',
		},
	},
	share_dir => {
		dist => 'share',
	},
	script_files => [
		'script/perldist_w'
	],
);

# To support our 899_prereq test, we have to wrap this.
unless ('Test::Prereq::Build' eq ref $builder) {
	my $answer = $builder->y_n('Do you want to install the minimal minicpan script?', 'n');

	if ($answer) {
		$builder->script_files(['script\minicpan_pdwix', 'script\perldist_w']);
		$builder->requires({
			'CPAN::Mini'        => '0.576',
			'CPAN::Mini::Devel' => '0.03',
			'File::Slurp'       => '9999.13',
			( %{$requires} )
		});
	}
}

$builder->create_build_script();
